File: examples/flutter/cases/tasks/task_move/task_move.agentic.case.dart
import '../../../core/agentic.case.dart';
import '../../../core/api.case.dart';
import '../../../core/shared/app_structural_contracts.dart';
import 'task_move.domain.case.dart';

typedef _TaskMoveHandler =
    Future<ApiResponse<TaskMoveOutput>> Function(TaskMoveInput input);

class TaskMoveAgentic extends BaseAgenticCase<TaskMoveInput, TaskMoveOutput> {
  TaskMoveAgentic(super.ctx);

  @override
  TaskMoveDomain domain() {
    return TaskMoveDomain();
  }

  @override
  AgenticDiscovery discovery() {
    return AgenticDiscovery(
      name: domainCaseName() ?? 'task_move',
      description:
          domainDescription() ??
          'Move an existing task card to another board column.',
      category: 'tasks',
      tags: const <String>['tasks', 'move', 'status'],
      aliases: const <String>[
        'move_task',
        'change_task_status',
        'advance_task',
      ],
      capabilities: const <String>['task_move', 'board_mutation'],
      intents: const <String>[
        'move a task',
        'change task status',
        'advance work on the board',
      ],
    );
  }

  @override
  AgenticExecutionContext context() {
    return const AgenticExecutionContext(
      requiresAuth: false,
      dependencies: <String>[
        'task_move.domain',
        'task_move.api',
        'task_list.agentic',
      ],
      preconditions: <String>[
        'A concrete taskId and a valid targetStatus are required.',
      ],
      constraints: <String>[
        'Use task_list first when the user refers to a task ambiguously.',
        'Execution must delegate to the canonical API surface.',
      ],
      notes: <String>[
        'Moving a task is a mutating action and should be confirmed by the host runtime.',
      ],
    );
  }

  @override
  AgenticPrompt prompt() {
    return const AgenticPrompt(
      purpose: 'Move an existing task to todo, doing, or done by task id.',
      whenToUse: <String>[
        'When the user explicitly wants to move an existing task card.',
        'When the user wants to update the progress status of known work.',
      ],
      whenNotToUse: <String>[
        'When the user wants to create a task.',
        'When the user has not provided enough information to identify the task.',
      ],
      constraints: <String>[
        'Do not invent a taskId.',
        'Require confirmation before mutating the board.',
      ],
      reasoningHints: <String>[
        'If the task is ambiguous, list current tasks first and ask the user to confirm the intended card.',
      ],
      expectedOutcome:
          'The updated task object with the requested target status persisted.',
    );
  }

  @override
  AgenticToolContract<TaskMoveInput, TaskMoveOutput> tool() {
    final inputSchema = domainInputSchema();
    final outputSchema = domainOutputSchema();

    if (inputSchema == null || outputSchema == null) {
      throw Exception('task_move.agentic requires domain schemas');
    }

    return AgenticToolContract<TaskMoveInput, TaskMoveOutput>(
      name: 'task_move',
      description: 'Move a task through the canonical API execution flow.',
      inputSchema: inputSchema,
      outputSchema: outputSchema,
      parseInput: (dynamic value) =>
          value is TaskMoveInput ? value : TaskMoveInput.fromJson(value),
      isMutating: true,
      requiresConfirmation: true,
      execute: (TaskMoveInput input, AgenticContext runtimeContext) async {
        final handler = _resolveHandler(runtimeContext);
        final result = await handler(input);

        if (!result.success || result.data == null) {
          throw toAppCaseError(result.error, 'task_move API failed');
        }

        return result.data!;
      },
    );
  }

  @override
  AgenticMcpContract mcp() {
    return const AgenticMcpContract(
      enabled: true,
      name: 'task_move',
      title: 'Move Task',
      description:
          'Move a task between board columns through the canonical APP task_move API flow.',
      metadata: <String, dynamic>{'category': 'tasks', 'mutating': true},
    );
  }

  @override
  AgenticRagContract rag() {
    return const AgenticRagContract(
      topics: <String>[
        'task_management',
        'status_transitions',
        'board_mutation',
      ],
      resources: <RagResource>[
        RagResource(
          kind: 'case',
          ref: 'tasks/task_move',
          description:
              'Canonical board mutation capability for changing task status.',
        ),
        RagResource(
          kind: 'case',
          ref: 'tasks/task_list',
          description:
              'Grounding capability used to identify persisted task ids before moving them.',
        ),
      ],
      hints: <String>[
        'Use task_list to ground ambiguous references before proposing or executing a move.',
        'Preserve explicit confirmation because task_move mutates persisted board state.',
      ],
      scope: 'project',
      mode: 'recommended',
    );
  }

  @override
  AgenticPolicy policy() {
    return const AgenticPolicy(
      requireConfirmation: true,
      riskLevel: 'medium',
      executionMode: 'manual-approval',
      limits: <String>[
        'Do not execute a move without explicit confirmation from the host runtime.',
      ],
    );
  }

  _TaskMoveHandler _resolveHandler(AgenticContext runtimeContext) {
    final tasksDomain = runtimeContext.cases?['tasks'];
    if (tasksDomain is! Map<String, dynamic>) {
      throw const AppCaseError(
        'INTERNAL',
        'task_move requires ctx.cases.tasks',
      );
    }

    final taskMove = tasksDomain['task_move'];
    if (taskMove is! Map<String, dynamic>) {
      throw const AppCaseError(
        'INTERNAL',
        'task_move requires ctx.cases.tasks.task_move',
      );
    }

    final handler = taskMove['apiHandler'];
    if (handler is _TaskMoveHandler) {
      return handler;
    }

    if (handler is Function) {
      return (TaskMoveInput input) async {
        final response = await Function.apply(handler, <dynamic>[input]);
        if (response is ApiResponse) {
          return ApiResponse<TaskMoveOutput>(
            success: response.success,
            data: response.data as TaskMoveOutput?,
            error: response.error,
            statusCode: response.statusCode,
          );
        }

        throw const AppCaseError(
          'INTERNAL',
          'task_move canonical API handler returned an invalid response',
        );
      };
    }

    throw const AppCaseError(
      'INTERNAL',
      'task_move requires a canonical API handler',
    );
  }

  @override
  Future<void> test() async {
    validateDefinition();

    final definition = this.definition();
    if (definition.tool.requiresConfirmation != true) {
      throw Exception('test: task_move tool must require confirmation');
    }

    if (definition.policy?.executionMode != 'manual-approval') {
      throw Exception('test: task_move should default to manual approval');
    }

    if ((definition.rag?.resources?.isEmpty ?? true)) {
      throw Exception('test: task_move should publish semantic RAG resources');
    }

    final example = examples().firstWhere(
      (AgenticExample<TaskMoveInput, TaskMoveOutput> item) =>
          item.name == 'move_todo_to_doing',
    );
    Future<ApiResponse<TaskMoveOutput>> successHandler(TaskMoveInput _) async {
      return ApiResponse.success(example.output);
    }

    Future<ApiResponse<TaskMoveOutput>> failureHandler(TaskMoveInput _) async {
      return ApiResponse.failure(
        const AppErrorData(
          code: 'NOT_FOUND',
          message: 'Task missing was not found',
        ),
      );
    }

    final result = await tool().execute(
      example.input,
      AgenticContext(
        correlationId: 'task-move-agentic-test',
        logger: ctx.logger,
        cases: <String, dynamic>{
          'tasks': <String, dynamic>{
            'task_move': <String, dynamic>{'apiHandler': successHandler},
          },
        },
      ),
    );

    if (result.task.status != example.output.task.status) {
      throw Exception('test: task_move tool must return the moved task');
    }

    Object? propagatedError;
    try {
      await tool().execute(
        example.input,
        AgenticContext(
          correlationId: 'task-move-agentic-failure-test',
          logger: ctx.logger,
          cases: <String, dynamic>{
            'tasks': <String, dynamic>{
              'task_move': <String, dynamic>{'apiHandler': failureHandler},
            },
          },
        ),
      );
    } catch (error) {
      propagatedError = error;
    }

    if (propagatedError is! AppCaseError) {
      throw Exception('test: task_move must propagate AppCaseError failures');
    }

    if (propagatedError.code != 'NOT_FOUND') {
      throw Exception('test: task_move must preserve NOT_FOUND from API');
    }
  }
}

File: examples/deno/cases/tasks/task_move/task_move.agentic.case.ts
import {
  type AgenticContext,
  type AgenticDiscovery,
  type AgenticExample,
  type AgenticExecutionContext,
  type AgenticMcpContract,
  type AgenticPolicy,
  type AgenticPrompt,
  type AgenticRagContract,
  type AgenticToolContract,
  BaseAgenticCase,
} from "../../../core/agentic.case";
import { type ApiResponse } from "../../../core/api.case";
import {
  AppCaseError,
  toAppCaseError,
} from "../../../core/shared/app_structural_contracts";
import {
  TaskMoveDomain,
  type TaskMoveInput,
  type TaskMoveOutput,
} from "./task_move.domain.case";

type ExpectedCasesMap = {
  tasks?: {
    task_move?: {
      api?: {
        handler(input: TaskMoveInput): Promise<ApiResponse<TaskMoveOutput>>;
      };
    };
  };
};

export class TaskMoveAgentic extends BaseAgenticCase<
  TaskMoveInput,
  TaskMoveOutput
> {
  protected domain(): TaskMoveDomain {
    return new TaskMoveDomain();
  }

  public discovery(): AgenticDiscovery {
    return {
      name: this.domainCaseName() ?? "task_move",
      description: this.domainDescription() ??
        "Move an existing task card to another board column.",
      category: "tasks",
      tags: ["tasks", "move", "status"],
      aliases: ["move_task", "change_task_status", "advance_task"],
      capabilities: ["task_move", "board_mutation"],
      intents: [
        "move a task",
        "change task status",
        "advance work on the board",
      ],
    };
  }

  public context(): AgenticExecutionContext {
    return {
      requiresAuth: false,
      dependencies: ["task_move.domain", "task_move.api", "task_list.agentic"],
      preconditions: [
        "A concrete taskId and a valid targetStatus are required.",
      ],
      constraints: [
        "Use task_list first when the user refers to a task ambiguously.",
        "Execution must delegate to the canonical API surface.",
      ],
      notes: [
        "Moving a task is a mutating action and should be confirmed by the host runtime.",
      ],
    };
  }

  public prompt(): AgenticPrompt {
    return {
      purpose: "Move an existing task to todo, doing, or done by task id.",
      whenToUse: [
        "When the user explicitly wants to move an existing task card.",
        "When the user wants to update the progress status of known work.",
      ],
      whenNotToUse: [
        "When the user wants to create a task.",
        "When the user has not provided enough information to identify the task.",
      ],
      constraints: [
        "Do not invent a taskId.",
        "Require confirmation before mutating the board.",
      ],
      reasoningHints: [
        "If the task is ambiguous, list current tasks first and ask the user to confirm the intended card.",
      ],
      expectedOutcome:
        "The updated task object with the requested target status persisted.",
    };
  }

  public tool(): AgenticToolContract<TaskMoveInput, TaskMoveOutput> {
    const inputSchema = this.domainInputSchema();
    const outputSchema = this.domainOutputSchema();

    if (!inputSchema || !outputSchema) {
      throw new Error("task_move.agentic requires domain schemas");
    }

    return {
      name: "task_move",
      description: "Move a task through the canonical API execution flow.",
      inputSchema,
      outputSchema,
      isMutating: true,
      requiresConfirmation: true,
      execute: async (input, ctx) => {
        const cases = ctx.cases as ExpectedCasesMap | undefined;
        const result = await cases?.tasks?.task_move?.api?.handler(input);

        if (!result?.success || !result.data) {
          throw toAppCaseError(result?.error, "task_move API failed");
        }

        return result.data;
      },
    };
  }

  public mcp(): AgenticMcpContract {
    return {
      enabled: true,
      name: "task_move",
      title: "Move Task",
      description:
        "Move a task between board columns through the canonical APP task_move API flow.",
      metadata: {
        category: "tasks",
        mutating: true,
      },
    };
  }

  public rag(): AgenticRagContract {
    return {
      topics: ["task_management", "status_transitions", "board_mutation"],
      resources: [
        {
          kind: "case",
          ref: "tasks/task_move",
          description:
            "Canonical board mutation capability for changing task status.",
        },
        {
          kind: "case",
          ref: "tasks/task_list",
          description:
            "Grounding capability used to identify persisted task ids before moving them.",
        },
      ],
      hints: [
        "Use task_list to ground ambiguous references before proposing or executing a move.",
        "Preserve explicit confirmation because task_move mutates persisted board state.",
      ],
      scope: "project",
      mode: "recommended",
    };
  }

  public policy(): AgenticPolicy {
    return {
      requireConfirmation: true,
      riskLevel: "medium",
      executionMode: "manual-approval",
      limits: [
        "Do not execute a move without explicit confirmation from the host runtime.",
      ],
    };
  }

  public examples(): AgenticExample<TaskMoveInput, TaskMoveOutput>[] {
    return super.examples();
  }

  public async test(): Promise<void> {
    this.validateDefinition();

    const definition = this.definition();
    if (!definition.tool.requiresConfirmation) {
      throw new Error("test: task_move tool must require confirmation");
    }

    if (definition.policy?.executionMode !== "manual-approval") {
      throw new Error("test: task_move should default to manual approval");
    }

    if (!definition.rag?.resources?.length) {
      throw new Error("test: task_move should publish semantic RAG resources");
    }

    const example = this.examples().find((item) =>
      item.name === "move_todo_to_doing"
    );
    if (!example) {
      throw new Error("test: task_move example must exist");
    }

    const mockCtx: AgenticContext = {
      correlationId: "task-move-agentic-test",
      logger: this.ctx.logger,
      cases: {
        tasks: {
          task_move: {
            api: {
              handler: async () => ({
                success: true,
                data: example.output,
              }),
            },
          },
        },
      },
    };

    const result = await this.tool().execute(example.input, mockCtx);
    if (result.task.status !== example.output.task.status) {
      throw new Error("test: task_move tool must return the moved task");
    }

    let propagatedError: unknown;
    try {
      await this.tool().execute(example.input, {
        correlationId: "task-move-agentic-failure-test",
        logger: this.ctx.logger,
        cases: {
          tasks: {
            task_move: {
              api: {
                handler: async () => ({
                  success: false,
                  error: {
                    code: "NOT_FOUND",
                    message: "Task missing was not found",
                  },
                }),
              },
            },
          },
        },
      });
    } catch (error: unknown) {
      propagatedError = error;
    }

    if (!(propagatedError instanceof AppCaseError)) {
      throw new Error("test: task_move must propagate AppCaseError failures");
    }

    if (propagatedError.code !== "NOT_FOUND") {
      throw new Error("test: task_move must preserve NOT_FOUND from API");
    }
  }
}
